File: 242.有效的字母异位词.py
#
# @lc app=leetcode.cn id=242 lang=python3
#
# [242] 有效的字母异位词
#

# @lc code=start
class Solution:
    def isAnagram(self,s:str, t:str)->bool:
        record=[0]*26
        # string 1 遍历
        for i in s:
            record[ord(i)-ord('a')]+=1
        for i in t:
            record[ord(i)-ord('a')]-=1
        
        # 判断
        for i in range(26):
            if record[i]!=0:
                return False
        return True
        # print(record)
        # return not all(record[i] for i in range(26)) # 这种写法不对！！

        # if len(s)!=len(t):
        #     return False
        # table=26*[0]
        # for i in s :
        #     index=ord(i)-ord('a')
        #     table[index]+=1
        # for i in t:
        #     index=ord(i)-ord('a')
        #     table[index]-=1
        #     if table[index]<0:
        #         return False
        # return True
'''
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        record = [0] * 26
        for i in s:
            #并不需要记住字符a的ASCII，只要求出一个相对数值就可以了
            record[ord(i) - ord("a")] += 1
        for i in t:
            record[ord(i) - ord("a")] -= 1
        for i in range(26):
            if record[i] != 0:
                #record数组如果有的元素不为零0，说明字符串s和t 一定是谁多了字符或者谁少了字符。
                return False
        return True
'''


# @lc code=end

File: 242.有效的字母异位词.cpp
/*
 * @lc app=leetcode.cn id=242 lang=cpp
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
class Solution {
public:
    bool isAnagram(string s, string t){
        if(s.length()!=t.length()){
            return false;
        }

        vector<int> record(26,0);//26个0

        for(int i=0; i<s.length();++i){
            int index=s[i]-'a';
            record[index]++;
        }

        for(int i=0;i<t.length();++i){
            int index=t[i]-'a';
            record[index]--;

            //减完要是负的 就不行

            if(record[index]<0){
                return false;
            }
        }
        return true;
    }
};

/*
class Solution {
public:
    bool isAnagram(string s, string t) {
        // 如果两个字符串的长度都不一致，那么肯定是无法成为字母异位词的
        if(s.length() != t.length()){

            // 直接返回 false
            return false;

        }

        // 让 a - z 这 26 个字母对应的下标变成 0 - 25 方便存到数组中
        // 比如 a 对应的索引就是 0 
        // b 对应的索引就是 1
        vector<int> table(26, 0);

        // 从头到尾遍历字符串 s
        for( int i = 0 ; i < s.length() ; i++){

            // 把访问的字符转换为整数的形式
            // 比如访问字母 a，那么 -'a' 之后就是 0，就是 a 对应的索引为 0 
            int index = s[i] - 'a';

            // 那么意味着这个字母出现的频次需要加 1
            table[index]++;

        }

        for( int i = 0 ; i < t.length() ; i++){

            // 把访问的字符转换为整数的形式
            // 比如访问字母 a，那么 -'a' 之后就是 0，就是 a 对应的索引为 0 
            int index = t[i] - 'a';

            // 那么意味着这个字母出现的频次需要减 1 
            table[index]--;

            // 如果说发现这个字母出现的频次小于了 0 
            // 说明 t 中出现了 s 中不曾用的字母
            if(table[index] < 0){

                // 那就不是字母异位词
                return false;

            }

        }

        // 否则，说明是字母异位词
        return true;
    }
};
*/
// @lc code=end

